sklearn/feature_selection/mutual_info_.py

```diff
@@ -0,0 +1,416 @@
+        with shape (n_features,) or array with indices of discrete features.
+        If 'auto', it is assigned to False for dense `X` and to True for
+        sparse `X`.
+    discrete_target : bool, default False
+        Whether to consider `y` as a discrete variable.
+    n_neighbors : int, default 3
+        Number of neighbors to use for MI estimation for continuous variables,
+        see [1]_ and [2]_. Higher values reduce variance of the estimation, but
+        could introduce a bias.
+    copy : bool, default True
+        Whether to make a copy of the given data. If set to False, the initial
+        data will be overwritten.
+    random_state : int seed, RandomState instance or None, default None
+        The seed of the pseudo random number generator for adding small noise
+        to continuous variables in order to remove repeated values.
+
+    Returns
+    -------
+    mi : ndarray, shape (n_features,)
+        Estimated mutual information between each feature and the target.
+        A negative value will be replaced by 0.
+
+    References
+    ----------
+    .. [1] A. Kraskov, H. Stogbauer and P. Grassberger, "Estimating mutual
+           information". Phys. Rev. E 69, 2004.
+    .. [2] B. C. Ross "Mutual Information between Discrete and Continuous
+           Data Sets". PLoS ONE 9(2), 2014.
+    """
+    X, y = check_X_y(X, y, accept_sparse='csc', y_numeric=not discrete_target)
+    n_samples, n_features = X.shape
+
+    if discrete_features == 'auto':
+        discrete_features = issparse(X)
+
+    if isinstance(discrete_features, bool):
+        discrete_mask = np.empty(n_features, dtype=bool)
+        discrete_mask.fill(discrete_features)
+    else:
+        discrete_features = np.asarray(discrete_features)
+        if discrete_features.dtype != 'bool':
+            discrete_mask = np.zeros(n_features, dtype=bool)
+            discrete_mask[discrete_features] = True
+        else:
+            discrete_mask = discrete_features
+
+    continuous_mask = ~discrete_mask
+    if np.any(continuous_mask) and issparse(X):
+        raise ValueError("Sparse matrix `X` can't have continuous features.")
+
+    if copy:
+        X = X.copy()
+
+    if not discrete_target and np.any(continuous_mask):
+        X[:, continuous_mask] = scale(X[:, continuous_mask],
+                                      with_mean=False, copy=False)
+
+    # Add small noise to continuous features as advised in Kraskov et. al.
+    rng = check_random_state(random_state)
+    if np.any(continuous_mask):
+        X = X.astype(float)
+        means = np.maximum(1, np.mean(np.abs(X[:, continuous_mask]), axis=0))
+        X[:, continuous_mask] += 1e-10 * means * rng.randn(
+                n_samples, np.sum(continuous_mask))
+
+    if not discrete_target:
+        y = scale(y, with_mean=False)
+        y += 1e-10 * np.maximum(1, np.mean(np.abs(y))) * rng.randn(n_samples)
+
+    mi = [_compute_mi(x, y, discrete_feature, discrete_target) for
```

Comment:
Umm, I'm not an expert in such kind of things, but it seems to me this part is negligible.